### ISSUE TITLE
Memory leakage with custom autograd_function 

### ISSUE THREAD

**Issue by adsb85** [mlverse/torch#403]
I am experiencing large memory leakage when using a custom built autograd_function. 
Specifically, in the example presented below, I create a custom 'linear' autograd_function, using the 'extending autograd vignette' example. [https://torch.mlverse.org/docs/articles/extending-autograd.html](url)

Over the course of training, the RAM consumed by the Rsession quickly explodes from ~700mb to 4-5Gb on CPUs. 
I do not experience the memory leakage when using the torch 'nn_linear' function.

Reproducible example provided below:

[memory_leakage_autograd_function.R.txt](https://github.com/mlverse/torch/files/5721425/memory_leakage_autograd_function.R.txt)

**Comment by dfalbel** [comment]
Thanks for reporting @adsb85 ! I will investigate

**Comment by dfalbel** [comment]
Hi @adsb85 I think this is related to #396 . It's seems that memory is leaking but actually, it's just that the R GC is lazy and was not called yet, so for example, if you add a `gc` call in your loop, things are going to work as expected.

PR #402 Will probably fix both issues.

**Comment by adsb85** [comment]
HI @dfalbel thanks for the quick response. I've tried adding a gc() call to the loop but unfortunately the Rsession continues to build up memory. 

I also tried clearing all variables and running a garbage collection, ie:

rm(list=ls()) 
gc()

But the Rsession continues to hold onto memory. The only way I can get it to release memory is by restarting the Rsession.  This behaviour seems to only occur when using custom autograd_functions. All autograd functions supplied by 'torch' do not have a memory leakage. 

PS thanks for brining torch to R 👍

**Comment by dfalbel** [comment]
Thanks @adsb85 ! I indeed found some leaks in the autograd_function implementation but AFAICT  they shouldn't lead to 5GB leakage neither I could reproduce the same behavior that you were seeing. 

Can you try the dev version with:

```
remotes::install_github("mlverse/torch")
```

and see if the problem still happens?

**Comment by adsb85** [comment]
Thanks again @dfalbel for following up. I am away from my computer during the holidays but will try your suggestion in using the dev version when I get back next week. 

For context, I only observe the ~5Gb leakage in the example that I provided if I run an absurdly large number of epochs during training. In the example that I provided I ran 2000 epochs (iterations) during training. After 1000 epochs I observe 2-3 Gbs of leakage, and after 2000 epochs I observe ~5Gb of leakage. If I only run <200 epochs then the leakage is not noticeable. 

That said, for larger scale problems, where the custom autograd function is performing many more operations and/or saving more tensors for the backward pass, then the leakage is more obvious after a smaller number of epochs.

**Comment by dfalbel** [comment]
Hi @adsb85,

THis version is now on CRAN. I'll close this issue but feel free to reopen if the problem persists.

**Comment by dfalbel** [comment]
This issue is still happening and unfortunately it's not detectable with memory leak detection tools.
i think the problem is that gradient disabling is applied per thread, and we need do it in each thread doing the computations.

**Comment by adsb85** [comment]
Hi @dfalbel  - just thought to re-up the memory leakage issue. I am currently running torch v0.6.0. However I am still experiencing large memory leak issues with custom autograd functions running on CPU (as per above). 

Thanks again!

**Comment by dfalbel** [comment]
Thanks @adsb85 for the heads up, sorry about that. It's on todo to fix that but I still can't give you a timeline. 
While in that I want to make sure to fix #732

**Comment by dfalbel** [comment]
@adsb85 and @rdinnager 

I found the potential leak and it seems it's been solved in #750 
Here's my test code. A few bytes might be shown but I almost sure they are false positives and related mostly to R's jit compiler caching / string pool caching. 

Please re-open this issue if you continue seeing the problem in the dev version.

```
f <- function() {
  linear <- autograd_function(
    forward = function(ctx, input, weight, bias = NULL) {
      ctx$save_for_backward(input = input, weight = weight, bias = bias)
      output <- input$mm(weight$t())
      if (!is.null(bias))
        output <- output + bias#$unsqueeze(1)$expand_as(output)
      
      output
    },
    backward = function(ctx, grad_output) {
      
      s <- ctx$saved_variables
      
      grads <- list(
        input = NULL,
        weight = NULL,
        bias = NULL
      )
      
      if (ctx$needs_input_grad$input)
        grads$input <- grad_output$mm(s$weight)
      
      if (ctx$needs_input_grad$weight)
        grads$weight <- grad_output$t()$mm(s$input)
      
      if (!is.null(s$bias) && ctx$needs_input_grad$bias)
        grads$bias <- grad_output$sum(dim = 1)
      
      grads
    }
  )
  
  # input dimensionality (number of input features)
  d_in <- 50
  # output dimensionality (number of predicted features)
  d_out <- 50
  # number of observations in training set
  n <- 100
  
  
  # create random data
  x <- torch_randn(n, d_in)
  y <- x +  torch_randn(n, d_out)
  
  w <- torch_randn(d_in, d_out, requires_grad = TRUE)
  b <- torch_randn(d_out, requires_grad = TRUE)
  
  learning_rate <- 0.1
  
  opt <- optim_adam(list(w, b))
  
  for (i in 1:10) {
    y_pred <- linear(x, w, b)
    loss <- nnf_mse_loss(y_pred, y)
    opt$zero_grad()
    loss$backward()
    opt$step()
  }
  
  invisible(NULL)
}

f()
gc(full = TRUE)
gc(full = TRUE)
first <- current <- lobstr::mem_used()
for (i in 1:5) {
  gc(full = TRUE)
  gc(full = TRUE)
  f()
  gc(full = TRUE)
  gc(full = TRUE)
  actual <- lobstr::mem_used()
  print(actual - current)
  current <- actual
}
cat("Total leaked memory:")
cat(current-first)
```

**Comment by rdinnager** [comment]
Awesome! Thanks very much for your work on this. I will give it a test with {styleganr} and let you know if the issue is resolved.

**Comment by rdinnager** [comment]
Just tested {styleganr} with this fix, and can confirm it works! Thanks very much, now I can resume development on it. I plan to release a generator-only package once I get it polished up (I will work on allowing model training for a future version).

**Comment by dfalbel** [comment]
Cool! Sorry for taking so long to fix it.
